File: cfg4/tests/CInterpolationTest.cc
/*
 * Copyright (c) 2019 Opticks Team. All Rights Reserved.
 *
 * This file is part of Opticks
 * (see https://bitbucket.org/simoncblyth/opticks).
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); 
 * you may not use this file except in compliance with the License.  
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, 
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  
 * See the License for the specific language governing permissions and 
 * limitations under the License.
 */

#include <cassert>
#include <climits>

#include "CFG4_BODY.hh"
#include "OPTICKS_LOG.hh"

// npy-
#include "NPY.hpp"


// okc-
#include "Opticks.hh"
#include "OpticksHub.hh"

//#include "GGeo.hh"
#include "GVector.hh"
#include "GGeoBase.hh"
#include "GMaterialLib.hh"
#include "GBndLib.hh"

// cfg4-
#include "CG4.hh"
#include "CMPT.hh"
#include "CMaterialBridge.hh"
#include "CSurfaceBridge.hh"

// g4-
#include "G4Material.hh"
#include "G4OpticalSurface.hh"



/**
CInterpolationTest
====================

The GPU analogue of this is oxrap-/tests/interpolationTest

::

    CInterpolationTest --nointerpol   ## identity check
    CInterpolationTest                ## interpol check


**/


const char* TMPDIR = "$TMP/cfg4/CInterpolationTest" ; 

int main(int argc, char** argv)
{
    OPTICKS_LOG(argc, argv);

    LOG(info) << argv[0] ;

    Opticks ok(argc, argv);
    OpticksHub hub(&ok) ;

    LOG(error) << "[ CG4::CG4" ; 
    CG4 g4(&hub);
    LOG(error) << "] CG4::CG4" ; 

    CMaterialBridge* mbr = g4.getMaterialBridge();
    CSurfaceBridge*  sbr = g4.getSurfaceBridge();
    //mbr->dump();

    GGeoBase* ggb = hub.getGGeoBase();
    GBndLib* blib = ggb->getBndLib(); 

    bool interpolate = ok.hasOpt("nointerpol") ? false : true ; 

    unsigned nl_interpolate = unsigned(Opticks::DOMAIN_HIGH) - unsigned(Opticks::DOMAIN_LOW) + 1u ; 


    // PROBABLY NOT NEEDED ANYMORE : DONE IN GBndLib::load 
    NPY<double>* tex = blib->createBuffer();   // zipping together the dynamic buffer from materials and surfaces
    unsigned ndim = tex->getDimensions() ;
    assert( ndim == 5 );

    unsigned ni = tex->getShape(0);   // ~123: number of bnd
    unsigned nj = tex->getShape(1);   //    4: omat/osur/isur/imat
    unsigned nk = tex->getShape(2);   //    2: prop groups
    unsigned nl = interpolate ? nl_interpolate : tex->getShape(3);   //   39: wl samples   OR   820 - 60 + 1 = 761
    unsigned nm = tex->getShape(4);   //    4: float4 props


    

    const char* name = interpolate ? 
             "CInterpolationTest_interpol.npy"
          :
             "CInterpolationTest_identity.npy"
          ;



    NPY<double>* out = NPY<double>::make(ni,nj,nk,nl,nm);
    out->fill(-1.f);  // align unset to -1.f

    LOG(info) 
       << " interpolate (control with option: --nointerpol) " << interpolate
       << " name " << name
       << " tex " << tex->getShapeString()
       << " out " << out->getShapeString()
       ;

    unsigned nb = blib->getNumBnd();
    assert( ni == nb );
    assert( nj == 4 && nm == 4);

    glm::vec4 boundary_domain = Opticks::getDefaultDomainSpec() ;

    double wlow = boundary_domain.x ; 
    double wstep = interpolate ? 1.0f : boundary_domain.z ;   // 1.0f OR 20.0f  (nanometer)

    LOG(info) << " wlow " << wlow 
              << " wstep " << wstep 
              << " nl " << nl 
              ;


    const char* mkeys_0 = "RINDEX,ABSLENGTH,RAYLEIGH,REEMISSIONPROB" ;
    const char* mkeys_1 = "GROUPVEL,,, " ;  // <-- without trailing space the split giving 3 not 4 


    // gathering an Opticks tex buffer  from G4 material and surface properties
    // as machinery and interpolation test 
    //
    //  cf with almost the reverse action done by 
    //       CSurLib::addProperties
    //             converts GGeo properties detect/absorb/reflect_specular/reflect_diffuse 
    //
    //       GSurfaceLib::createStandardSurface
    //
    //      into G4 props
    //                  REFLECTIVITY 
    //                  EFFICIENCY (when sensor)


    LOG(info) << " nb " << nb ; 
    for(unsigned i=0 ; i < ni ; i++)
    {
        guint4 bnd = blib->getBnd(i);

        unsigned omat = bnd.x ; 
        unsigned osur = bnd.y ; 
        unsigned isur = bnd.z ; 
        unsigned imat = bnd.w ;   // these are zero based indices, UINT_MAX for unset 

        LOG(info) 
            << " i "    << std::setw(3) << i
            << " omat " << std::setw(3) << omat
            << " osur " << std::setw(3) << osur
            << " isur " << std::setw(3) << isur
            << " imat " << std::setw(3) << imat
            ;

        const G4Material* om = mbr->getG4Material(omat);
        const G4Material* im = mbr->getG4Material(imat);
        assert(om) ; 
        assert(im) ; 

        const G4OpticalSurface* os = osur == UINT_MAX ? NULL : sbr->getG4Surface(osur) ;
        const G4OpticalSurface* is = isur == UINT_MAX ? NULL : sbr->getG4Surface(isur) ;

        CMPT* ompt = new CMPT(om->GetMaterialPropertiesTable(), om->GetName().c_str()); 
        CMPT* impt = new CMPT(im->GetMaterialPropertiesTable(), im->GetName().c_str()); 
        CMPT* ospt = os == NULL ? NULL : new CMPT(os->GetMaterialPropertiesTable(), os->GetName().c_str());
        CMPT* ispt = is == NULL ? NULL : new CMPT(is->GetMaterialPropertiesTable(), is->GetName().c_str());
         
        for(unsigned k = 0 ; k < nk ; k++)
        { 
            const char* mkeys = k == 0 ? mkeys_0 : mkeys_1 ; 

            unsigned om_offset = out->getValueIndex(i, GBndLib::OMAT, k ) ;
            unsigned os_offset = out->getValueIndex(i, GBndLib::OSUR, k ) ;
            unsigned is_offset = out->getValueIndex(i, GBndLib::ISUR, k ) ;
            unsigned im_offset = out->getValueIndex(i, GBndLib::IMAT, k ) ;

            ompt->sample(out, om_offset, mkeys, wlow, wstep, nl );  
            impt->sample(out, im_offset, mkeys, wlow, wstep, nl );  

            if(k == 0)  // nothing in 2nd group for surfaces yet ???
            {
                if(ospt) 
                {
                    bool ospecular = os->GetFinish() == polished ;
                    ospt->sampleSurf(out, os_offset, wlow, wstep, nl, ospecular );  
                } 
                if(ispt) 
                {
                    bool ispecular = is->GetFinish() == polished ;
                    ispt->sampleSurf(out, is_offset, wlow, wstep, nl, ispecular  );  
                }
            }
        }


        LOG(info) << std::setw(5) << i 
                  << "(" 
                  << std::setw(2) << omat << ","
                  << std::setw(2) << ( osur == UINT_MAX ? -1 : (int)osur ) << ","
                  << std::setw(2) << ( isur == UINT_MAX ? -1 : (int)isur ) << ","
                  << std::setw(2) << imat 
                  << ")" 
                  << std::setw(60) << blib->shortname(bnd) 
                  << " om " << std::setw(30) << ( om ? om->GetName() : "-" ) 
                  << " im " << std::setw(30) << ( im ? im->GetName() : "-" ) 
                  ; 

    } 

    out->save(TMPDIR,name);

    return 0 ; 
}

File: cfg4/tests/CInterpolationTest.py
#!/usr/bin/env python
#
# Copyright (c) 2019 Opticks Team. All Rights Reserved.
#
# This file is part of Opticks
# (see https://bitbucket.org/simoncblyth/opticks).
#
# Licensed under the Apache License, Version 2.0 (the "License"); 
# you may not use this file except in compliance with the License.  
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software 
# distributed under the License is distributed on an "AS IS" BASIS, 
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  
# See the License for the specific language governing permissions and 
# limitations under the License.
#

"""

::

   ipython -i  CInterpolationTest.py
   ipython -i  CInterpolationTest.py -- --nointerpol


"""
import os, sys, logging, numpy as np
log = logging.getLogger(__name__)

from opticks.ana.base import opticks_main 
from opticks.ana.proplib import PropLib

np.set_printoptions(precision=3, suppress=True)



def check_exists(xpath, msg):
    if not os.path.exists(xpath):
        log.info(" NO SUCH PATH %s run \"%s\" " % (xpath, msg) )
        sys.exit(0)


def load(ok):
    if ok.interpol:
        path = "$TMP/InterpolationTest/CInterpolationTest_interpol.npy" 
        x_shape = (123, 4, 2, 761, 4)
        msg = "CInterpolationTest"
    else:
        path = "$TMP/InterpolationTest/CInterpolationTest_identity.npy" 
        x_shape = (123, 4, 2, 39, 4)
        msg = "CInterpolationTest --nointerpol"
    pass
    xpath = os.path.expandvars(path)
    check_exists(xpath, msg)
    c = np.load(xpath)

    log.info("load %s %r " % (xpath, c.shape))
    assert c.shape == x_shape, "shape mismatch expect %r found %r " % (x_shape, c.shape)
    return  c


class CFProp(object): 
    """
    ::

        In [23]: cf.t.shape
        Out[23]: (123, 4, 2, 39, 4)


    """
    def __init__(self, ok):
        blib = PropLib("GBndLib")
        names = blib.names
        t = np.load(os.path.expandvars("$IDPATH/GBndLib/GBndLib.npy"))
        c = load(ok)

        assert len(t) == len(names)
        assert len(t) == len(c) 
        n = len(t)

        self.shape = t.shape 
        self.names = names
        self.t = t
        self.c = c
        self.n = n

        self.consistency_check("t",t)
   


    def consistency_check(self,key,x):
        log.info(key)
        chk = {}
        for i in range(self.n):
            self.index = i 
            bnd = self.names[i]
            omat,osur,isur,imat = bnd.split("/")

            



    def __getitem__(self, sli):
        self.sli = sli 
        return self

    def __call__(self, arg):
        if type(arg) is int:
            name = self.names[arg]
        elif type(arg) is str:
            name = arg
        elif type(arg) is slice:
            return map(lambda name:self(name), self.names[arg])
        else:
            assert 0, (type(arg), "unexpected type")
        pass
        return self.check_bnd(name)

    def check_bnd(self, name):
        omat,osur,isur,imat = name.split("/")



if __name__ == '__main__':

    ok = opticks_main()

    blib = PropLib("GBndLib")
    names = blib.names
    n = len(names)

    nam = np.zeros( (n,4 ), dtype="|S64")
    for i,name in enumerate(names):nam[i] = name.split("/")




    cf = CFProp(ok)



if 0:
    for i in range(n):

        name = names[i]

        g4_omat = np.all( t[i,blib.B_OMAT,0] == c[i,blib.B_OMAT,0] )
        g4_imat = np.all( t[i,blib.B_IMAT,0] == c[i,blib.B_IMAT,0] )


        if omat in g4:
            assert g4[omat] == g4_omat
        else:
            g4[omat] = g4_omat

        if imat in g4:
            assert g4[imat] == g4_imat
        else:
            g4[imat] = g4_imat
 

        if len(osur) > 0:
            g4_osur = np.all( t[i,blib.B_OSUR,0] == c[i,blib.B_OSUR,0] )

            if osur in g4:
                assert g4[osur] == g4_osur
            else:
                g4[osur] = g4_osur
        else:
            g4_osur = None
        pass

     
        if len(isur) > 0:
            g4_isur = np.all( t[i,blib.B_ISUR,0] == c[i,blib.B_ISUR,0] )

            if isur in g4:
                assert g4[isur] == g4_isur
            else:
                g4[isur] = g4_isur
        else:
            g4_isur = None
        pass


        if g4_omat == False:
           if not omat in g4b:
               g4b[omat] = []
           g4b[omat].append( (i,blib.B_OMAT,0) )  

        if g4_osur == False:
           if not osur in g4b:
               g4b[osur] = []
           g4b[osur].append( (i,blib.B_OSUR,0) )  

        if g4_isur == False:
           if not isur in g4b:
               g4b[isur] = []
           g4b[isur].append( (i,blib.B_ISUR,0) )  

        if g4_imat == False:
           if not imat in g4b:
               g4b[imat] = []
           g4b[imat].append( (i,blib.B_IMAT,0) )  
 

        print "%4d omat %25s imat %25s         g4_omat %7s g4_imat %7s " % (  i, omat, imat,  g4_omat, g4_imat )  

        #if len(isur) > 0 or len(osur) > 0:
        #    print "%4d osur %35s isur %35s         ok_osur %7s ok_isur %7s      g4_osur %7s g4_isur %7s " % (  i, osur, isur, ok_osur, ok_isur,  g4_osur, g4_isur )  
              
 


    pass


    print "g4", g4

    for k,v in g4b.items():
        if len(v) > 0:print k, str(v)





